webnn/clamp.https.any.js

```diff
@@ -0,0 +1,214 @@
+// META: title=test WebNN API clamp operation
+// META: global=window,dedicatedworker
+// META: script=./resources/utils.js
+// META: timeout=long
+
+'use strict';
+
+// https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-clamp
+
+const testClamp = async (operandType, syncFlag, inputValue, inputShape, expected, options) => {
+  const x = builder.input('x', {type: operandType, dimensions: inputShape});
+  const y = builder.clamp(x, options);
+  const TestTypedArray = TypedArrayDict[operandType];
+  const inputs = {'x': new TestTypedArray(inputValue)};
+  const outputs = {'y': new TestTypedArray(sizeOfShape(inputShape))};
+  let graph;
+
+  if (syncFlag) {
+    graph = builder.build({y});
+    context.compute(graph, inputs, outputs);
+  } else {
+    graph = await builder.buildAsync({y});
+    await context.computeAsync(graph, inputs, outputs);
+  }
+
+  const expectedData = expected instanceof Array ? new TestTypedArray(expected) : expected[operandType];
```

Comment:
Why not just use `TestTypedArray(expected)` if the expected results are double precision results generated by webnn-baseline? At line 14, the input is `new TestTypedArray(inputValue)`. I suppose they should be aligned.

Reply:
Hi @huningxin, current webnn-baseline could generate `float64` baseline (expected data) by feeding `float64` input data which @wchao1115 and CG also agreed on, while @fdwr last commented that it's unfair to simply casting `float64` baseline to be `float32` baseline since using actual casted `float32` input data would make precision loss  from original **float64** input data. 
So I workaround webnn-baseline to support generate **float32** baseline, please review this generation code [here](https://github.com/BruceDai/webnn-baseline-1/pull/6/files), thanks. 

Also ping @wchao1115 and @fdwr  to have a review this baseline generation code, thanks.

Reply:
It matters that the JS baseline and the C++ implementation start with the same value (it doesn't *really* matter whether casting is done in the generated data directly or whether it's done at run time on the input data). Otherwise, if the JS implementation receives the more accurate `3.14159265` input, but C++ implementation sees an already truncated value of `3.14159274`, then data loss has already occurred outside its control, and we'd be measuring the C++ implementation against an unfair standard (which would be like telling a function to reverse the characters in a string "CAT" <-> "TAC" but chopping off the last character "CA", yet still expecting it to return "TAC" instead of "AC").

The approach above works for float32, but for testing float16, I think we need to write some casting helper function (since I don't see a Float16Array in ES). The test cases below should include some float16 too eventually, or at least a placeholder enum for the data type.

```
    // 1D
    [[24], expectedDefault, DataType.float16],
    [[24], expectedDefault, DataType.float32],
```

Reply:
> For float16 (which is probably a later target), you probably need some casting helper function, since I don't see a Float16Array in ES.

@wchao1115 commented previously on https://github.com/webmachinelearning/webnn/issues/127 :

"Although the lack of FP16 support in the ArrayBufferView could possibly be worked around by just using Uint16Array to appropriately offset the tensor data and deal with the float-casting inside the webnn implementation itself, it could still be confusing to some at the model builder level."

 There is a workaround by 森建 on https://esdiscuss.org/topic/float16array :

"WebGL 2.0 supports half-precision float (16bit float) by default.
But now, we must use the following dirty hack using Uint16Array."


```js
// ref: http://stackoverflow.com/questions/32633585/how-do-you-convert-to-half-floats-in-javascript
var toHalf = (function() {

   var floatView = new Float32Array(1);
   var int32View = new Int32Array(floatView.buffer);

   /* This method is faster than the OpenEXR implementation (very often
    * used, eg. in Ogre), with the additional benefit of rounding, inspired
    * by James Tursa?s half-precision code. */
   return function toHalf(val) {

     floatView[0] = val;
     var x = int32View[0];

     var bits = (x >> 16) & 0x8000; /* Get the sign */
     var m = (x >> 12) & 0x07ff; /* Keep one extra bit for rounding */
     var e = (x >> 23) & 0xff; /* Using int is faster here */

     /* If zero, or denormal, or exponent underflows too much for a denormal
      * half, return signed zero. */
     if (e < 103) {
       return bits;
     }

     /* If NaN, return NaN. If Inf or exponent overflow, return Inf. */
     if (e > 142) {
       bits |= 0x7c00;
       /* If exponent was 0xff and one mantissa bit was set, it means NaN,
        * not Inf, so make sure we set one mantissa bit too. */
       bits |= ((e == 255) ? 0 : 1) && (x & 0x007fffff);
       return bits;
     }

     /* If exponent underflows but not too much, return a denormal */
     if (e < 113) {
       m |= 0x0800;
       /* Extra rounding may overflow and set mantissa to 0 and exponent
        * to 1, which is OK. */
       bits |= (m >> (114 - e)) + ((m >> (113 - e)) & 1);
       return bits;
     }

     bits |= ((e - 112) << 10) | (m >> 1);
     /* Extra rounding. An overflow will set mantissa to 0 and increment
      * the exponent, which is OK. */
     bits += m & 1;
     return bits;
   };

}());

var tex = new Uint16Array(4);
tex[0] = toHalf(0.5);
tex[1] = toHalf(1);
tex[2] = toHalf(123);
tex[3] = toHalf(-13);
```